contributors/design-proposals/core-metrics-pipeline.md

```diff
@@ -0,0 +1,280 @@
+type CpuUsage struct {  
+  // The time at which these Metrics were updated.  
+  Timestamp metav1.Time  
+  // Average CPU usage rate over sample window (across all cores), in "nano cores".  
+  // The "core" unit represents nanoseconds of CPU time consumed per second.  
+  // For example, 5 nanocores means the process averaged 5 nanoseconds 
+  // of cpu time per second during the sample window.
+  UsageRateNanoCores *uint64  
+  // Cumulative CPU usage (sum of all cores) since object creation.  
+  AggregateUsageCoreNanoSeconds *uint64   
+}  
+
+// MemoryUsage holds statistics about the quantity of memory consumed  
+type MemoryUsage struct {  
+  // The time at which these metrics were updated.  
+  Timestamp metav1.Time  
+  // The amount of working set memory. This includes recently accessed memory,  
+  // dirty memory, and kernel memory.  
+  WorkingSetBytes *uint64  
+}  
+
+// VolumeUsage holds statistics about the quantity of disk resources consumed for a volume  
+type VolumeUsage struct {  
+  // Embedded FilesystemUsage  
+  FilesystemUsage  
+  // Name is the name given to the Volume  
+  Name string  
+}  
+
+// FilesystemUsage holds statistics about the quantity of disk resources consumed  
+type FilesystemUsage struct {  
+  // The time at which these metrics were updated.  
+  Timestamp metav1.Time  
+  // The name filesystme on which resources are consumed.  This must uniquely identify the filesystem.
+  Filesystem string  
+  // UsedBytes represents the disk space consumed on the filesystem, in bytes.  
+  UsedBytes *uint64  
+  // InodesUsed represents the inodes consumed on the filesystem  
+  InodesUsed *uint64  
+}  
+```
+
+### On-Demand Design
+Interface:  
+The interface for exposing these metrics within the kubelet contains a method for getting this datastructure.  This method contains a "recency" parameter which specifies how recently the metrics must have been computed.  Kubelet components which require very up-to-date metrics (eviction, for example), use very low values.  Other components use higher values.  
+
+Implementation:  
+To keep performance bounded while still offering metrics "On-Demand", all calls to get metrics are cached, and a minimum recency is established to prevent repeated metrics computation.  Before computing new metrics, the previous metrics are checked to see if they meet the recency requirements of the caller.  If the age of the metrics meet the recency requirements, then the cached metrics are returned.  If not, then new metrics are computed and cached.  
+In the case where some metrics are not able to be computed instantly (time-averaged metrics, for example), then they must be recomputed on the "minimum recency" interval so that they always meet recency requirements.
+
+## Implementation Plan
+@dashpole will create an interface over cAdvisor that provides core metrics.  
+@dashpole will modify volume metrics collection so that it re-uses vendored cAdvisor libraries.   
```

Comment:
this is much smaller of an issue that I first thought.  Not worthy of including in a proposal